packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -575,6 +642,51 @@ export class PermissionController<
     });
   }
 
+  /**
+   * Gets the permission specification corresponding to the given permission
+   * type and target name. Throws an error if the target name does not
+   * correspond to a permission, or if the specification is not of the
+   * given permission type.
+   *
+   * @template Type - The type of the permission specification to get.
+   * @param permissionType - The type of the permission specification to get.
+   * @param targetName - The name of the permission whose specification to get.
+   * @param requestingOrigin - The origin of the requesting subject, if any.
+   * Will be added to any thrown errors.
+   * @returns The specification object corresponding to the given type and
+   * target name.
+   */
+  private getTypedPermissionSpecification<Type extends PermissionType>(
+    permissionType: Type,
+    targetName: string,
+    requestingOrigin?: string,
+  ): ControllerPermissionSpecification & { permissionType: Type } {
+    const failureError =
+      permissionType === PermissionType.RestrictedMethod
+        ? methodNotFound(
+            targetName,
+            requestingOrigin ? { origin: requestingOrigin } : undefined,
+          )
+        : new EndowmentPermissionDoesNotExistError(
+            targetName,
+            requestingOrigin,
+          );
+
+    const targetKey = this.getTargetKey(targetName);
+    if (!targetKey) {
+      throw failureError;
+    }
+
+    const specification = this.getPermissionSpecification(targetKey);
+    if (specification.permissionType !== permissionType) {
+      throw failureError;
+    }
+
+    return specification as unknown as ControllerPermissionSpecification & {
+      permissionType: Type;
+    };
+  }
```

Comment:
It took some doing, but I was able to do this without casting by using a type predicate:

```suggestion
    if (!this.matchesType(specification, permissionType)) {
      throw failureError;
    }

    return specification;
  }

  /**
   * Checks that the specification has the expected permission type.
   *
   * @param specification - The specification to check.
   * @param expectedType - The expected permission type.
   * @template Specification - The specification to check.
   * @template Type - The expected permission type.
   * @returns Whether or not the specification is of the expected type.
   */
  private matchesType<
    Specification extends PermissionSpecificationConstraint,
    Type extends PermissionType,
  >(
    specification: Specification,
    expectedType: Type,
  ): specification is Specification & {
    permissionType: Type;
  } {
    return specification.permissionType === expectedType;
  }
```